kernel/src/process.rs

```diff
@@ -1143,43 +1290,190 @@ impl<C: Chip> ProcessType for Process<'_, C> {
+            return GenericSyscallReturnValue::AllowReadWriteFailure(
+                ErrorCode::FAIL,
+                buf_start_addr,
+                size,
+            );
+        }
+
+        // A process is allowed to pass any pointer if the slice
+        // length is 0, as to revoke kernel access to a memory region
+        // without granting access to another one
+        let new_slice = if size == 0 {
+            // Clippy complains that we're deferencing a pointer in a
+            // public and safe function here. While we are not
+            // deferencing the pointer here, we pass it along to an
+            // unsafe function, which is as dangerous (as it is likely
+            // to be deferenced down the line).
+            //
+            // Relevant discussion:
+            // https://github.com/rust-lang/rust-clippy/issues/3045
+            //
+            // It should be fine to ignore the lint here, as a slice
+            // of length 0 will never allow dereferencing any memory
+            // in a safe manner.
+            unsafe { ReadWriteAppSlice::new(buf_start_addr, 0, self.appid()) }
+        } else if self.in_app_owned_memory(buf_start_addr, size) {
+            // TODO: Check for buffer aliasing here
+
+            // Valid slice, we need to adjust the app's watermark
+            // note: in_app_owned_memory ensures this offset does not wrap
+            let buf_end_addr = buf_start_addr.wrapping_add(size);
+            let new_water_mark = max(self.allow_high_water_mark.get(), buf_end_addr);
+            self.allow_high_water_mark.set(new_water_mark);
+
+            // Clippy complains that we're deferencing a pointer in a
+            // public and safe function here. While we are not
+            // deferencing the pointer here, we pass it along to an
+            // unsafe function, which is as dangerous (as it is likely
+            // to be deferenced down the line).
+            //
+            // Relevant discussion:
+            // https://github.com/rust-lang/rust-clippy/issues/3045
+            //
+            // It should be fine to ignore the lint here, as long as
+            // we make sure that we're pointing towards userspace
+            // memory (verified using `in_app_owned_memory`) and
+            // respect alignment and other constraints of the Rust
+            // references created by ReadWriteAppSlice.
+            unsafe { ReadWriteAppSlice::new(buf_start_addr, size, self.appid()) }
+        } else {
+            return GenericSyscallReturnValue::AllowReadWriteFailure(
+                ErrorCode::INVAL,
+                buf_start_addr,
+                size,
+            );
+        };
+
+        let allow_result = driver_invoc_closure(new_slice);
+
+        // TODO: Check whether the capsule either made the allow fail
+        // and returned the new slice, or it made the allow succeed
+        // and returned the expected slice from the allow table
```

Comment:
TODO

Reply:
Yes. Enforcing allow semantics is going to be an important part of the kernel work before release.

Reply:
marking resolved as this will be handled post-merge